library/Booster/Pattern/UnifiedMatcher.hs

```diff
@@ -0,0 +1,839 @@
+    , mSubsorts :: SortTable
+    }
+
+type SortTable = Map SortName (Set SortName)
+
+match :: MatchType -> StateT MatchState (Except MatchResult) ()
+match matchType = do
+    queue <- gets mQueue
+    mapQueue <- gets mMapQueue
+    case queue of
+        Seq.Empty -> case mapQueue of
+            Seq.Empty -> checkIndeterminate -- done
+            (term1, term2) :<| rest -> do
+                modify $ \s -> s{mMapQueue = rest}
+                match1 matchType term1 term2
+                match matchType
+        (term1, term2) :<| rest -> do
+            modify $ \s -> s{mQueue = rest}
+            match1 matchType term1 term2
+            match matchType
+
+checkIndeterminate :: StateT MatchState (Except MatchResult) ()
+checkIndeterminate = do
+    indeterminate <- gets mIndeterminate
+    unless (null indeterminate) . lift $
+        throwE (MatchIndeterminate $ NE.fromList indeterminate)
+match1 ::
+    MatchType ->
+    Term ->
+    Term ->
+    StateT MatchState (Except MatchResult) ()
+{- FOURMOLU_DISABLE -}
+match1 Rewrite (AndTerm t1a t1b)                          t2@AndTerm{}                               = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 Eval    t1@AndTerm{}                               t2@AndTerm{}                               = addIndeterminate t1 t2
+match1 _       (AndTerm t1a t1b)                          t2@DomainValue{}                           = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@Injection{}                             = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@KMap{}                                  = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@KList{}                                 = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@KSet{}                                  = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@ConsApplication{}                       = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 _       (AndTerm t1a t1b)                          t2@FunctionApplication{}                   = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 Rewrite (AndTerm t1a t1b)                          t2@Var{}                                   = enqueueRegularProblem t1a t2 >> enqueueRegularProblem t1b t2
+match1 Eval    t1@AndTerm{}                               t2@Var{}                                   = addIndeterminate t1 t2
+match1 Rewrite t1@DomainValue{}                           (AndTerm t2a t2b)                          = enqueueRegularProblem t1 t2a >> enqueueRegularProblem t1 t2b
```

Comment:
I like how this looks! I'd suggest a minor style change: let's add comments separating the groups of clauses for the different `Term` patterns: 

```suggestion
-- DomainValue
match1 Rewrite t1@DomainValue{} (AndTerm t2a t2b) = enqueueRegularProblem t1 t2a >> enqueueRegularProblem t1 t2b
```